### PR TITLE
DataObjectBuilder also for mutations

### PR DESCRIPTION
**Pull request by xaviapa**

### CODE REVIEW

File: tests/DataObjectBuilderTest.php [hunk 1]
```diff
@@ -35,15 +37,15 @@ public function testWhenDataIsAnEmptyArray()
...
 
         $expectedDataObject = [
-            'search' => [],
+            'search' => new QueryCollection([]),
         ];
         $this->assertEquals($expectedDataObject, $dataObject);
     }
 
-    public function testWhenDataHasAnItem()
+    public function testWhenDataHasAnQueryItem()
```
Comment (joskfg) [inline tests/DataObjectBuilderTest.php]: *AnQuery

File: src/Client.php [hunk 2]
```diff
@@ -17,10 +17,12 @@ public function __construct(ClientInterface $httpClient, ResponseBuilder $respon
     }
 
     /**
+     * @param mixed $variables Variables can be in array or MutationObject formats.
+     *
      * @throws \UnexpectedValueException When response body is not a valid json
      * @throws \RuntimeException         When there are transfer errors
      */
-    public function query(string $query, array $variables = null): Response
+    public function query(string $query, $variables = null): Response
```
Comment (rccrdpccl) [inline src/Client.php]: Wouldn't it be better to leave this as "raw query" and use another method for mutation purposes, for example `mutate` or `mutation`?

File: src/Mutation/FilteredCollection.php [hunk 3]
```diff
@@ -4,14 +4,14 @@
...
-class FilteredCollection implements MutationObject, \JsonSerializable
+class FilteredCollection implements MutationObject, \IteratorAggregate, \JsonSerializable
 {
     use MutationObjectHandler;
 
     /**
      * @var array<Item>
      */
-    protected $arguments = [];
+    public $arguments = [];
```
Comment (rccrdpccl) [inline src/Mutation/FilteredCollection.php]: Why public?

File: src/Mutation/Item.php [hunk 4]
```diff
@@ -8,10 +8,19 @@ class Item implements MutationObject, \JsonSerializable
 {
     use MutationObjectHandler;
 
-    private $arguments;
-
+    /**
+     * @var array
+     */
+    public $arguments;
```
Comment (rccrdpccl) [inline src/Mutation/Item.php]: Why public?

**Review by rccrdpccl — CHANGES_REQUESTED** [review]

File: src/Mutation/FilteredCollection.php (hunk 3 above)
Comment (xaviapa) [reply src/Mutation/FilteredCollection.php]: It's the only way the Iterator from FilteredCollection works. It does not work with a getter.

File: src/Client.php (hunk 2 above)
Comment (xaviapa) [reply src/Client.php]: The only difference would be that the mutate(9 can receive also a Mutation object. Is it worth?

File: src/Mutation/FilteredCollection.php (hunk 3 above)
Comment (rccrdpccl) [reply src/Mutation/FilteredCollection.php]: What exactly is not working?

File: src/Mutation/FilteredCollection.php [hunk 5]
```diff
@@ -47,6 +47,33 @@ public function set(array $data): void
...
+                $current = $this->current();
+                if ($current instanceof Item) {
+                    return false;
+                }
+
+                if (is_array($current)) {
+                    return true;
+                }
+
+                foreach ($current->arguments as $argument) {
```
Comment (rccrdpccl) [inline src/Mutation/FilteredCollection.php]: If this is the only reason to make it public, we should encapsulate this logic into a method `hasChild` or something like this

File: src/Mutation/FilteredCollection.php [hunk 6]
```diff
@@ -47,6 +47,46 @@ public function set(array $data): void
...
     }
 
+    public function getIterator()
+    {
+        return new \RecursiveIteratorIterator($this->getRecursiveArrayIterator());
+    }
+
+    public function getRecursiveArrayIterator(): \RecursiveArrayIterator
+    {
+        return new class($this->arguments) extends \RecursiveArrayIterator {
```
Comment (joskfg) [inline src/Mutation/FilteredCollection.php]: We may extract this

File: src/Mutation/FilteredCollection.php [hunk 7]
```diff
@@ -47,6 +47,46 @@ public function set(array $data): void
...
+                if (is_array($current)) {
+                    return true;
+                }
+
+                return $current->hasChildren();
+            }
+
+            public function getChildren()
+            {
+                return parent::current()->getRecursiveArrayIterator();
```
Comment (joskfg) [inline src/Mutation/FilteredCollection.php]: We could use something like:

```
return parent::current()->getIterator()->getInnerIterator()
```

With this approach we will have just one public method to obtain an iterator.

File: src/Client.php (hunk 2 above)
Comment (rccrdpccl) [reply src/Client.php]: Well, in my view mutate can *only* receive Mutation object. Right now query and variables expect the "raw" format: it is quite weird to pass half "raw" (query syntax for mutation) and half automatic (mutation object). Ideally the mutation object has the name of the mutation itself, so no raw query is needed. Am I missing anything here?

**Review by rccrdpccl — APPROVED** [review]

**Review by jeurrutia — APPROVED** [review]

**Review by rccrdpccl — APPROVED** [review]

**Review by joskfg — APPROVED** [review]

**Pull request merged**